items: migrate to TRX

lahm86 opened the pull request:
#### Checklist

- [x] I have read the [coding conventions](https://github.com/LostArtefacts/TRX/blob/master/CONTRIBUTING.md#coding-conventions)
- [x] I have added a changelog entry about what my pull request accomplishes, or it is an internal change
- [x] I have added a readme entry about my new feature or OG bug fix, or it is a different change

#### Description

I'm going to do an incremental PR for this move, as it's a bit more involved. I'll add a comment with each increment for what to check, then once approved will move to the next stage and re-request review.

lahm86 committed:
tr1/misc: use Item_Get

This replaces most cases of direct g_Items access with Item_Get.

lahm86 committed:
tr2/misc: use Item_Get

This replaces msot cases of direct g_Items access with Item_Get.

lahm86 committed:
tr2/lara: simplify M16 drawing test

This removes duplication for testing if Lara is using the M16 for her
drawing routines.

lahm86 committed:
items: introduce Item_GetIndex

This provides a common function to retrieve an item index.

lahm86 commented:
The first stage replaces the majority of `g_Items[idx]` calls with `Item_Get`. `items.c` in both games remains untouched as I want to try to move some common routines that use the array out to TRX, so will do that in following PRs.

The only logical changes at the moment are as follows:
- adjusting the check for Lara locking on to a new target in TR1
- simplifying a check for Lara's drawing in TR2 (while she's holding or using the M16)

lahm86 committed:
items: remove global item level count

This removes the global level item count and makes it private in the
items module. It also adds a convenience function for creating an item
and incrementing this count e.g. for pods and statues.

lahm86 committed:
items: remove next/prev active item globals

This moves the next/prev item globals to TRX and makes them private.

lahm86 commented:
Second stage removes the global _level_ item count in favour of a getter, and adds a utility for such things as pods and statues to create an item _and_ increment the level count. This count is independent of the actual total item count, as we need to retain it for savegames.

Good things to test in various save/load states (TombATI saves would be good as well to check):
- the injected PSX mummy in Khamoon
- pods
- centaur statues
- Thor's hammer
- Skidoo drivers
- The dragon

aredfan reviewed (approved):
LGTM👍

rr- commented on src/tr2/game/lara/misc.c:
```diff
@@ -1803,3 +1803,14 @@ void Lara_Extinguish(void)
+    const ITEM *const item = Item_Get(g_Lara.weapon_item);
+    return item->current_anim_state == 0 || item->current_anim_state == 2
+        || item->current_anim_state == 4;
+}
```
Can we add a description to what this function does?

lahm86 replied:
Thanks, done. Added it to the header. It's sort of an awkward function but as the code was duplicated in four places, I felt it warranted separating out.

rr- reviewed (approved):
Thank you.

lahm86 committed:
items: move storage to TRX

This moves item storage to TRX, along with some functions that directly
manipulate the array such as killing and adding active items.

lahm86 committed:
level: read items in TRX

This moves item reading to TRX.

lahm86 committed:
level: load objects and items in TRX

This moves the object setup and item initialisation calls to TRX in a
common routine.

lahm86 commented:
The last three commits above should wrap-up this PR for now. Item storage and reading is now fully in TRX, and I've adjusted object setup and item initialisation to be called commonly from TRX as well, because of the need for this to happen after injections and in the defined order.

A general test here would be good to ensure items activate (enemies, traps, doors etc) and that they are killed. And a sanity check that item state persists over save/load would be good. Thanks.

lahm86 committed:
level/types: remove redundant level item count

This is no longer required for injection processing.

rr- commented on src/libtrx/game/items.c:
```diff
@@ -60,6 +92,148 @@ int16_t Item_CreateLevelItem(void)
+#else
+    if (item_num < m_LevelItemCount) {
+        item->flags |= IF_KILLED;
+    }
```
I wonder why we're doing this check in TR2+. I mean, why would the game treat dynamically created objects any different?

lahm86 replied:
Yes, it seems a bit odd, but I figured safer for now to keep it as-is. I'll need to study elsewhere to see if the flag is used in different ways possibly first.

rr- commented on src/libtrx/game/level/common.c:
```diff
@@ -822,3 +823,16 @@ void Level_LoadPalettes(LEVEL_INFO *const info)
+void Level_LoadObjectsAndItems(void)
+{
+    // Object and item setup/initialisation must take place after injections
+    // have been processed. We also need to use a cached item count as
```
Let's avoid 'we' and use more robot wording.

rr- reviewed (approved):
It's unfortunate that there are places that need to mess with `m_PrevItemActive` (and need `Item_SetPrevActive`) but it is what it is.

lahm86 commented:
> It's unfortunate that there are places that need to mess with `m_PrevItemActive` (and need `Item_SetPrevActive`) but it is what it is.

Yes, it is a shame. I _believe_ this is only needed for the body bag feature; I'm not sure if we want to retain that as it seems only required for performance reasons, but that's a separate task to investigate anyway. Would much prefer not to expose this property if we could.
